### DISCUSSION TITLE
Embedded fonts in PDF?

### DISCUSSION THREAD

**Discussion by afdev82** [prawnpdf/prawn#1241] [Q&A] [answered]
Just a simple question: are the built-in fonts (like the default, Helvetica) or the external ones (eg. Avenir) embedded in the PDF?
In other words, do I need that font in my system to render the text when I open the generated PDF?

I ask this because we are having problems with a software for a digital printer that doesn't recognize the font (probably I'll have more details on it soon).
In my Ubuntu installation I don't have Helvetica or Avenir installed, but I see the text without problems, so I thought that the font was embedded in the generated PDF.

**Comment by pointlessone** [comment, marked as answer]
No, default fonts are not embedded. Adobe was cunning enough to convince everyone that having a default set of fonts pre-installed with every PDF renderer would save a lot of space/bandwidth. The fonts were pre-installed with every Adobe Acrobat Reader and for a while it was pretty much the only option out there. But now there are many renderers out there. Starting with PDF 1.5 special treatment of the default 14 fonts is deprecated but for compatibility reasons renderers should still do it anyway.

I'm not sure about the licensing of the fonts. I'd expect printer manufacturers be able to get them from Adobe if they support PDF natively.

Prawn still generates PDFs around v 1.4 that are expected to have this special treatment for default fonts but for some features we set higher format version. Maybe your digital printer prematurely dropped support for the default PDF fonts or they have not licensed the fonts. Maybe it's something different.

**Reply by afdev82** [reply]
Thank you for the fast reply!

I'm reading on many sites that embedding fonts in PDF is recommended in order to have the same content appearance everywhere (that should be the goal of the PDF format in fact). If I need always the fonts in my system the format is not really "Portable".
Using other software (eg. Illustrator) I can embed the fonts without any problem (eg. Arial or Helvetica).
I found [a page](https://helpx.adobe.com/acrobat/using/pdf-fonts.html) on the Adobe website where they explain how embedding fonts work, so I don't think they are against this.

Is there a way to embed the fonts anyway with Prawn?
In the README I read that Prawn has:

> Support for both PDF builtin fonts as well as embedded TrueType fonts

What does it mean then "embedded TrueType fonts"?

The generated PDF is v1.4 and Acrobat shows that in fact the Helvetica font is not embedded.
On my PC running Ubuntu I can see that with my PDF reader the Helvetica font gets replaced by Arial.
If I open the PDF generated with Illustrator with embedded fonts, I can see the Helvetica also on my PC.

I think to solve the problem I should either:

- ask the printing manufacturer to support the default fonts
- install the missing font on the printing machine system
- use a font available on the printing machine system (eg. Arial), but to do that I have to buy the font (I assume I can't just copy the font on the server)

**Reply by pointlessone** [reply]
You absolutely can embed fonts with Prawn. Please refer to `font` and `font_families` documentation. There are examples in the docs. Also see External Fonts section in the [manual](https://prawnpdf.org/manual.pdf).

I can't really advice on how to deal with the printer. TBH, I doubt your printer actually supports PDF natively (very few do). Most likely printer driver translates PDF into something like PCL. You most likely can install free fonts that look pretty similar to the default PDF fonts on your printing machine. E.g. Ghostscript has all the fonts required by PDF spec. I have no idea how to configure the driver to pick those fonts up, though.

Speaking of Ghostscript, I think you can use it to actually embed default fonts. Might be useful if you have to deal with user-submitted PDFs. I can't help with the actual incantation though.

**Reply by afdev82** [reply]
Hi, thank you for your information.

Now I understand more about fonts in Prawn / PDF.
I saw that other fonts we are using (Verdana or Avenir) are embedded in the PDF, but not the built-in fonts.
As far as I know the printing machine software imports the PDF file and if some font is missing just throws an error.
I'm pretty flexible in fact, because I generate the PDFs and I can choose which font to use.
We tested with the printer and with the embedded fonts is working fine.

Thank you very much!
